common/version/semver.go

```diff
@@ -0,0 +1,147 @@
+package version
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+)
+
+// Semver represents a semantic version.
+type Semver struct {
+	major  int
+	minor  int
+	patch  int
+	errata string
+}
+
+// NewSemver creates a new Semver instance.
+func NewSemver(major int, minor int, patch int, errata string) (*Semver, error) {
+	if major < 0 {
+		return nil, fmt.Errorf("major version must be non-negative")
+	}
+	if minor < 0 {
+		return nil, fmt.Errorf("minor version must be non-negative")
+	}
+	if patch < 0 {
+		return nil, fmt.Errorf("patch version must be non-negative")
+	}
+
+	return &Semver{
+		major:  major,
+		minor:  minor,
+		patch:  patch,
+		errata: errata,
+	}, nil
+}
+
+// Parses a semantic version string and returns a Semver instance.
```

Comment:
Consider describing here how this handles errata and extra, at a high level

Reply:
```
// Parses a semantic version string and returns a Semver instance.
//
// Requires the string to have the following format: X.Y.Z[-errata], where X, Y, and Z are
// non-negative integers, and errata is an optional string (which may contain dots). Note that if
// errata is present, it must be preceded by a hyphen, e.g. "1.2.3-alpha.1".
func SemverFromString(versionStr string) (*Semver, error) {
```